README.md
![skycoin logo](https://user-images.githubusercontent.com/26845312/32426705-d95cb988-c281-11e7-9463-a3fce8076a72.png)

# Skycoin

[![Build Status](https://travis-ci.com/skycoin/skycoin.svg)](https://travis-ci.com/skycoin/skycoin)
[![GoDoc](https://godoc.org/github.com/skycoin/skycoin?status.svg)](https://godoc.org/github.com/skycoin/skycoin)
[![Go Report Card](https://goreportcard.com/badge/github.com/skycoin/skycoin)](https://goreportcard.com/report/github.com/skycoin/skycoin)

Skycoin is a next-generation cryptocurrency.

Skycoin was written from scratch and designed over four years to realize the
ideal of Bitcoin and represents the apex of cryptocurrency design.
Skycoin is not designed to add features to Bitcoin,
but rather improves Bitcoin by increasing simplicity,
security and stripping out everything non-essential.

Some people have hyped the Skycoin Project as leading into "Bitcoin 3.0".
The coin itself is not "Bitcoin 3.0",
but is rather "Bitcoin 1.0". Bitcoin is a prototype crypto-coin.
Skycoin was designed to be what Bitcoin would look like if it were built from
scratch, to remedy the rough edges in the Bitcoin design.

- no duplicate coin-base outputs
- enforced checks for hash collisions
- simple deterministic wallets
- no transaction malleability
- no signature malleability
- removal of the scripting language
- CoinJoin and normal transactions are indistinguishable
- elimination of edge-cases that prevent independent node implementations
- <=10 second transaction times
- elimination of the need for mining to achieve blockchain consensus

## Links

* [skycoin.com](https://www.skycoin.com)
* [Skycoin Blog](https://www.skycoin.com/blog)
* [Skycoin Docs](https://www.skycoin.com/docs)
* [Skycoin Blockchain Explorer](https://explorer.skycoin.com)
* [Skycoin Development Telegram Channel](https://t.me/skycoindev)
* [Skycoin Github Wiki](https://github.com/skycoin/skycoin/wiki)

## Table of Contents

<!-- MarkdownTOC levels="1,2,3,4,5" autolink="true" bracket="round" -->

- [Changelog](#changelog)
- [Installation](#installation)
	- [Go 1.14+ Installation and Setup](#go-1.14-installation-and-setup)
	- [Go get skycoin](#go-get-skycoin)
	- [Run Skycoin from the command line](#run-skycoin-from-the-command-line)
	- [Show Skycoin node options](#show-skycoin-node-options)
	- [Run Skycoin with options](#run-skycoin-with-options)
	- [Docker image](#docker-image)
	- [Building your own images](#building-your-own-images)
	- [Development image](#development-image)
- [API Documentation](#api-documentation)
	- [REST API](#rest-api)
	- [Skycoin command line interface](#skycoin-command-line-interface)
- [Integrating Skycoin with your application](#integrating-skycoin-with-your-application)
- [Contributing a node to the network](#contributing-a-node-to-the-network)
- [Creating a new coin](#creating-a-new-coin)
- [Daemon CLI Options](#daemon-cli-options)
- [URI Specification](#uri-specification)
- [Wire protocol user agent](#wire-protocol-user-agent)
- [Offline transaction signing](#offline-transaction-signing)
- [Deploy a public Skycoin API node with HTTPS](#deploy-a-public-skycoin-api-node-with-https)
- [Development](#development)
	- [Modules](#modules)
	- [Client libraries](#client-libraries)
	- [Running Tests](#running-tests)
	- [Running Integration Tests](#running-integration-tests)
		- [Stable Integration Tests](#stable-integration-tests)
		- [Live Integration Tests](#live-integration-tests)
		- [Debugging Integration Tests](#debugging-integration-tests)
		- [Update golden files in integration testdata](#update-golden-files-in-integration-testdata)
	- [Test coverage](#test-coverage)
		- [Test coverage for the live node](#test-coverage-for-the-live-node)
	- [Formatting](#formatting)
	- [Code Linting](#code-linting)
	- [Profiling](#profiling)
	- [Fuzzing](#fuzzing)
		- [base58](#base58)
		- [encoder](#encoder)
	- [Dependencies](#dependencies)
		- [Rules](#rules)
		- [Management](#management)
	- [Configuration Modes](#configuration-modes)
		- [Development Desktop Client Mode](#development-desktop-client-mode)
		- [Server Daemon Mode](#server-daemon-mode)
		- [Electron Desktop Client Mode](#electron-desktop-client-mode)
		- [Standalone Desktop Client Mode](#standalone-desktop-client-mode)
	- [Wallet GUI Development](#wallet-gui-development)
		- [Translations](#translations)
	- [Releases](#releases)
		- [Update the version](#update-the-version)
		- [Pre-release testing](#pre-release-testing)
		- [Creating release builds](#creating-release-builds)
		- [Release signing](#release-signing)
- [Responsible Disclosure](#responsible-disclosure)

<!-- /MarkdownTOC -->

## Changelog

[CHANGELOG.md](CHANGELOG.md)

## Installation

Skycoin supports go1.14+.

### Go 1.14+ Installation and Setup

[Golang 1.14+ Installation/Setup](./INSTALLATION.md)

### Go get skycoin

```sh
$ go get github.com/skycoin/skycoin/cmd/...
```

This will download `github.com/skycoin/skycoin` to `$GOPATH/src/github.com/skycoin/skycoin`.

You can also clone the repo directly with `git clone https://github.com/skycoin/skycoin`,
but it must be cloned to this path: `$GOPATH/src/github.com/skycoin/skycoin`.

### Run Skycoin from the command line

```sh
$ cd $GOPATH/src/github.com/skycoin/skycoin
$ make run-client
```

### Show Skycoin node options

```sh
$ cd $GOPATH/src/github.com/skycoin/skycoin
$ make run-help
```

### Run Skycoin with options

Example:

```sh
$ cd $GOPATH/src/github.com/skycoin/skycoin
$ make ARGS="--launch-browser=false -data-dir=/custom/path" run
```

### Docker image

This is the quickest way to start using Skycoin using Docker.

```sh
$ docker volume create skycoin-data
$ docker volume create skycoin-wallet
$ docker run -ti --rm \
    -v skycoin-data:/data/.skycoin \
    -v skycoin-wallet:/wallet \
    -p 6000:6000 \
    -p 6420:6420 \
    skycoin/skycoin
```

This image has a `skycoin` user for the skycoin daemon to run, with UID and GID 10000.
When you mount the volumes, the container will change their owner, so you
must be aware that if you are mounting an existing host folder any content you
have there will be own by 10000.

The container will run with some default options, but you can change them
by just appending flags at the end of the `docker run` command. The following
example will show you the available options.

```sh
$ docker run --rm skycoin/skycoin -help
```

Access the dashboard: [http://localhost:6420](http://localhost:6420).

Access the API: [http://localhost:6420/version](http://localhost:6420/version).

### Building your own images

[Building your own images](docker/images/mainnet/README.md).

### Development image

The [skycoin/skycoindev-cli docker image](docker/images/dev-cli/README.md) is provided in order to make
easy to start developing Skycoin. It comes with the compiler, linters, debugger
and the vim editor among other tools.

The [skycoin/skycoindev-dind docker image](docker/images/dev-docker/README.md) comes with docker installed
and all tools available on `skycoin/skycoindev-cli:develop` docker image.

Also, the [skycoin/skycoindev-vscode docker image](docker/images/dev-vscode/README.md) is provided
to facilitate the setup of the development process with [Visual Studio Code](https://code.visualstudio.com)
and useful tools included in `skycoin/skycoindev-cli`.

## API Documentation

### REST API

[REST API](src/api/README.md).

### Skycoin command line interface

[CLI command API](cmd/skycoin-cli/README.md).

## Integrating Skycoin with your application

[Skycoin Integration Documentation](INTEGRATION.md)

## Contributing a node to the network

Add your node's `ip:port` to the [peers.txt](peers.txt) file.
This file will be periodically uploaded to https://downloads.skycoin.com/blockchain/peers.txt
and used to seed client with peers.

*Note*: Do not add Skywire nodes to `peers.txt`.
Only add Skycoin nodes with high uptime and a static IP address (such as a Skycoin node hosted on a VPS).

## Creating a new coin

See the [newcoin tool README](./cmd/newcoin/README.md)

## Daemon CLI Options

See the [Skycoin Daemon CLI options](./cmd/skycoin/README.md)

## URI Specification

Skycoin URIs obey the same rules as specified in Bitcoin's [BIP21](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki).
They use the same fields, except with the addition of an optional `hours` parameter, specifying the coin hours.

Example Skycoin URIs:

* `skycoin:2hYbwYudg34AjkJJCRVRcMeqSWHUixjkfwY`
* `skycoin:2hYbwYudg34AjkJJCRVRcMeqSWHUixjkfwY?amount=123.456&hours=70`
* `skycoin:2hYbwYudg34AjkJJCRVRcMeqSWHUixjkfwY?amount=123.456&hours=70&label=friend&message=Birthday%20Gift`

Additonally, if no `skycoin:` prefix is present when parsing, the string may be treated as an address:

* `2hYbwYudg34AjkJJCRVRcMeqSWHUixjkfwY`

However, do not use this URI in QR codes displayed to the user, because the address can't be disambiguated from other Skyfiber coins.

## Wire protocol user agent

[Wire protocol user agent description](https://github.com/skycoin/skycoin/wiki/Wire-protocol-user-agent)

## Offline transaction signing

Before doing the offline transaction signing, we need to have the unsigned transaction created. Using the `skycoin-cli` tool to create an unsigned transaction in hot wallet, and copy the hex encoded transaction to the computer where the cold wallet is installed. Then use the `skycoin-cli` tool to sign it offline.

### Create an unsigned transaction

The `skycoin-cli` tool replys on the APIs of the `skycoin node`, hence we have to start the node before running the tool.

 Go to the project root and run:

```bash
$ ./run-client.sh -launch-browser=false
```

Once the node is started, we could use the following command to create an unsigned transaction.

```bash
$ skycoin-cli createRawTransactionV2 $WALLET_FILE $RECIPIENT_ADDRESS $AMOUNT --unsign
```

> Note: Don't forget the `--unsign` flag, otherwise it would try to sign the transaction.

<details>
 <summary>View Output</summary>

```json
b700000000e6b869f570e2bfebff1b4d7e7c9e86885dbc34d6de988da6ff998e7acd7e6e14010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000007531184ad0afeebbff2049b855e0921329cb1cb74d769ac57c057c9c8bd2b6810100000000ed5ea2ca4fe9b4560409b50c5bf7cb39b6c5ff6e50690f00000000000000000000000000
```

</details>

Copy and save the generated transaction string. We will sign it with a cold wallet offline in the next section.

### Sign the transaction

The `skycoin node` needs to have the most recently `DB` so that the user would not lose much coin hours when signing the transaction. We could copy the full synchronized `data.db` from the hot wallet to the computer where the cold wallet is installed. And place it in `$HOME/.skycoin/data.db`. Then start the node with the network disabled.

```bash
$ ./run-client.sh -launch-browser=false -disable-networking
```

Run the following command to sign the transaction:

```bash
$ skycoin-cli signTransaction $RAW_TRANSACTION
```

The `$RAW_TRANSACTION` is the transaction string that we generated in the hot wallet.

If the cold wallet is encrypted, you will be prompted to enter the password to sign the transaction.

<details>
 <summary>View Output</summary>

```json
b700000000e6b869f570e2bfebff1b4d7e7c9e86885dbc34d6de988da6ff998e7acd7e6e14010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000007531184ad0afeebbff2049b855e0921329cb1cb74d769ac57c057c9c8bd2b6810100000000ed5ea2ca4fe9b4560409b50c5bf7cb39b6c5ff6e50690f00000000000000000000000000
```

</details>

Once the transaction is signed, we could copy and save the signed transaction string and broadcast it in the hot wallet.

```bash
$ skycoin-cli broadcastTransaction $SIGNED_RAW_TRANSACTION
```

A transaction id would be returned and you can check it in the [explorer](https://explorer.skycoin.com).

## Deploy a public Skycoin API node with HTTPS

We recommend using [caddy server](https://caddyserver.com/) to deploy a public Skycoin API node on a
Linux server. The public API node should have the `HTTPS` support, which could be handled automatically
by the `caddy server`. But we need to have a domain and create a DNS record to bind the server ip address
to it.

Suppose we're going to deploy a Skycoin API node on `apitest.skycoin.com`, and we have already bound
the server's IP to it. Follow the steps below to complete the deployment.

### Install and run a skycoin api node

```bash
# Create a skycoin folder so that the files could be isolated
$ mkdir $HOME/skycoin && cd $HOME/skycoin
# Download the skycoin binary file
$ wget https://downloads.skycoin.com/wallet/skycoin-0.26.0-gui-standalone-linux-x64.tar.gz
$ tar -zxvf skycoin-0.26.0-gui-standalone-linux-x64.tar.gz
$ cd skycoin-0.26.0-gui-standalone-linux-x64
$ ./skycoin -web-interface-port=6420 -host-whitelist=$DOMAIN_NAME -enable-api-sets="READ,TXN"
```

> Note: we should running the `skycoin` node with `-host-whitelist` flag, otherwise it would
> throw `403 Forbidden` error.

### Install the caddy server

```bash
# Create a caddy folder
$ mkdir $HOME/caddy && cd $HOME/caddy
# Download the caddy server binary file
$ wget https://github.com/caddyserver/caddy/releases/download/v1.0.4/caddy_v1.0.4_linux_amd64.tar.gz
$ tar -zxvf caddy_v1.0.4_linux_amd64.tar.gz
$ cd caddy_v1.0.4_linux_amd64
```

The `caddy` tool would be exist in the folder, let's create a `Caddyfile` to define the reverse proxy
rules now.

```bash
cat <<EOF >Caddyfile
apitest.skycoin.com {
   proxy / localhost:6420 {
      transparent
   }
}
EOF
```

Then run the caddy server

```bash
$ ./caddy
```

You will be prompted to enter an email address to receive the notifications from let's Encrypt.
That's all about the deployment, check the https://apitest.skycoin.com/api/v1/version to see if
the Skycoin API node is working correctly.

## Development

We have two branches: `master` and `develop`.

`develop` is the default branch and will have the latest code.

`master` will always be equal to the current stable release on the website, and should correspond with the latest release tag.

### Modules

* `api` - REST API interface
* `cipher` - cryptographic library (key generation, addresses, hashes)
* `cipher/base58` - Base58 encoding
* `cipher/encoder` - reflect-based deterministic runtime binary encoder
* `cipher/encrypt` - at-rest data encryption (chacha20poly1305+scrypt)
* `cipher/go-bip39` - BIP-39 seed generation
* `cli` - CLI library
* `coin` - blockchain data structures (blocks, transactions, unspent outputs)
* `daemon` - top-level application manager, combining all components (networking, database, wallets)
* `daemon/gnet` - networking library
* `daemon/pex` - peer management
* `params` - configurable transaction verification parameters
* `readable` - JSON-encodable representations of internal structures
* `skycoin` - core application initialization and configuration
* `testutil` - testing utility methods
* `transaction` - methods for creating transactions
* `util` - miscellaneous utilities
* `visor` - top-level blockchain database layer
* `visor/blockdb` - low-level blockchain database layer
* `visor/historydb` - low-level blockchain database layer for historical blockchain metadata
* `wallet` - wallet file management

### Client libraries

Skycoin implements client libraries which export core functionality for usage from
other programming languages.

* [libskycoin C client library and SWIG interface](https://github.com/skycoin/libskycoin)
* [skycoin-lite: Javascript and mobile bindings](https://github.com/skycoin/skycoin-lite)

### Running Tests

```sh
$ make test
```

### Running Integration Tests

There are integration tests for the CLI and HTTP API interfaces. They have two
run modes, "stable" and "live".

The stable integration tests will use a skycoin daemon
whose blockchain is synced to a specific point and has networking disabled so that the internal
state does not change.

The live integration tests should be run against a synced or syncing node with networking enabled.

#### Stable Integration Tests

```sh
$ make integration-test-stable
```

or

```sh
$ ./ci-scripts/integration-test-stable.sh -v -w
```

The `-w` option, run wallet integrations tests.

The `-v` option, show verbose logs.

#### Live Integration Tests

The live integration tests run against a live runnning skycoin node, so before running the test, we
need to start a skycoin node:

```sh
$ ./run-daemon.sh
```

After the skycoin node is up, run the following command to start the live tests:

```sh
$ make integration-test-live
```

The above command will run all tests except the wallet-related tests. To run wallet tests, we
need to manually specify a wallet file, and it must have at least `2 coins` and `256 coinhours`,
it also must have been loaded by the node.

We can specify the wallet by setting two environment variables:

* `API_WALLET_ID`, which is the filename (without path), that is loaded by the daemon to test against.
  This is the `"id"` field in API requests. It is used by the API integration tests.
  The wallet directory that the daemon uses can be controlled with the `-wallet-dir` option.
* `CLI_WALLET_FILE`, which is the filename (with path), to be used by the CLI integration tests

If the wallet is encrypted, also set `WALLET_PASSWORD`.

Example of running the daemon with settings for integration tests:

```sh
$ export API_WALLET_ID="$valid_wallet_filename"
$ export CLI_WALLET_FILE="$HOME/.skycoin/wallets/$valid_wallet_filename"
$ export WALLET_PASSWORD="$wallet_password"
$ make run-integration-test-live
```

Then run the tests with the following command:

```sh
$ make integration-test-live-wallet
```

There are two other live integration test modes for CSRF disabled and networking disabled.

To run the CSRF disabled tests:

```sh
$ export API_WALLET_ID="$valid_wallet_filename"
$ export CLI_WALLET_FILE="$HOME/.skycoin/wallets/$valid_wallet_filename"
$ export WALLET_PASSWORD="$wallet_password"
$ make run-integration-test-live-disable-csrf
```

```sh
$ make integration-test-live-disable-csrf
```

To run the networking disabled tests, which require a live wallet:

```sh
$ export API_WALLET_ID="$valid_wallet_filename"
$ export CLI_WALLET_FILE="$HOME/.skycoin/wallets/$valid_wallet_filename"
$ export WALLET_PASSWORD="$wallet_password"
$ make run-integration-test-live-disable-networking
```

Then run the tests with the following command:

```sh
$ make integration-test-live-wallet
```

#### Debugging Integration Tests

Run specific test case:

It's annoying and a waste of time to run all tests to see if the test we real care
is working correctly. There's an option: `-r`, which can be used to run specific test case.
For example: if we only want to test `TestStableAddressBalance` and see the result, we can run:

```sh
$ ./ci-scripts/integration-test-stable.sh -v -r TestStableAddressBalance
```

#### Update golden files in integration testdata

Golden files are expected data responses from the CLI or HTTP API saved to disk.
When the tests are run, their output is compared to the golden files.

To update golden files, use the provided `make` command:

```sh
$ make update-golden-files
```

We can also update a specific test case's golden file with the `-r` option.
For example:
```sh
$ ./ci-scripts/integration-test-stable.sh -v -u -r TestStableAddressBalance
```

### Test coverage

Coverage is automatically generated for `make test` and integration tests run against a stable node.
This includes integration test coverage. The coverage output files are placed in `coverage/`.

To merge coverage from all tests into a single HTML file for viewing:

```sh
$ make check
$ make merge-coverage
```

Then open `coverage/all-coverage.html` in the browser.

#### Test coverage for the live node

Some tests can only be run with a live node, for example wallet spending tests.
To generate coverage for this, build and run the skycoin node in test mode before running the live integration tests.

In one shell:

```sh
$ make run-integration-test-live-cover
```

In another shell:

```sh
$ make integration-test-live
```

After the tests have run, CTRL-C to exit the process from the first shell.
A coverage file will be generated at `coverage/skycoin-live.coverage.out`.

Merge the coverage with `make merge-coverage` then open the `coverage/all-coverage.html` file to view it,
or generate the HTML coverage in isolation with `go tool cover -html`

### Formatting

All `.go` source files should be formatted `goimports`.  You can do this with:

```sh
$ make format
```

### Code Linting

Install prerequisites:

```sh
$ make install-linters
```

Run linters:

```sh
$ make lint
```

### Profiling

A full CPU profile of the program from start to finish can be obtained by running the node with the `-profile-cpu` flag.
Once the node terminates, a profile file is written to `-profile-cpu-file` (defaults to `cpu.prof`).
This profile can be analyzed with

```sh
$ go tool pprof cpu.prof
```

The HTTP interface for obtaining more profiling data or obtaining data while running can be enabled with `-http-prof`.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

- Add `POST /api/v1/wallet/scan` API to scan wallet addresses ahead, default scan number is `20`.
- Add `CLI walletScanAddresses` command to scan wallet addresses ahead.
- Add `GET /api/v2/transactions` API to get transactions with pagination.
- Add `-max-incoming-connection` flag to control the maximum allowed incoming connections.

### Fixed

### Changed

- Add `src_txid` to verbose transaction inputs, affected apis: `/api/v1/transaction`, `/api/v1/transactions` `/api/v1/block` `/api/v1/blocks` `/api/v1/pendingTxs`
- Add `calculated_hours` to `/api/v1/uxout` and `/api/v1/address_uxouts`.
- Update DB version only when DB structure is changed.
- Add go version of go1.14 support
- Change `coinhours` burn factor to 1 percent.

### Removed

## [0.27.0] - 2019-11-26

### Added

- Add `createRawTransactionV2` CLI command, which calls the API of `/wallet/transaction` to create the transaction and can create then unsigned transaction. Once the API's performance issue has been fixed, we will replace the `createRawTransaction` with it.
- Add `signTransaction` CLI command to sign transaction.
- Do windows electron builds by travis and abandon the appveyor
- Migrate `skycoin.net` to `skycoin.com`
- Migrate project path to `SkycoinProject/skycoin`
- Use transaction history when scanning wallet addresses, instead of the current address balance
- Document the daemon's CLI options
- Add the ability to save transaction notes
- Add CLI `encodeJsonTransaction` command to retrieve raw transaction given its JSON representation
- Add `package bip44`, implementing the bip44 spec https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki
- Codesign daemon and standalone binaries
- Add a guided method for entering the seeds in the GUI
- Add new wallet type `collection` for wallets that are an arbitrary collection of private keys, rather than generated from a seed
- Add new wallet type `bip44` for hierarchical deterministic (HD) wallets obeying the bip44 protocol.
  The Skycoin bip44 `coin` number is `8000`.
  `bip44` wallets avoid address reuse, generating a new change address for each transaction.
  Affects APIs are `POST /api/v1/wallet`, `GET /api/v1/wallets`, `GET /api/v1/wallet`, `POST /api/v1/wallet/seed` and `POST /api/v1/wallet/recover`.
  Refer to the [API documentation](./src/api/README.md) for API changes.
  `bip44` wallets support bip39 "seed passphrases".
  More details are explained in https://github.com/skycoin/skycoin/wiki/Wallet-File-Formats-and-Types
- `cli walletCreate` support for `bip44` wallets added
- Add `bip44_coin` field to `GET /api/v1/health` `fiber` params
- Add the "bulk send" option to the GUI advanced form
- Add `cli walletKeyExport` command to export `xpub`, `xprv`, `pub` and `prv` key from a `bip44` wallet
- Add `xpub` type wallet, which can generate addresses without exposing private keys
- Add `block_publisher` flag to `/api/v1/health`
- Add `no_broadcast` option to `POST /api/v1/injectTransaction` to disable broadcast of the transaction. The transaction will only be added to the local pool.
- Add `cli distributeGenesis` command to split the genesis block into distribution addresses when setting up a new fiber coin

### Fixed

- #7, #162, corrupted file in ~/.skycoin/data/ dir makes the desktop wallet show ERROR #1.
- #87, can not run web gui from skycoin docker image.
- #2287 A `Content-Type` with a `charset` specified, for example `application/json; charset=utf-8`, will not return an HTTP 415 error anymore
- Fix `fiber.toml` transaction verification parameters ignored by newcoin
- #2373 Fix and clean-up further panics with various `skycoin-cli` commands (lastBlocks, checkdb) which were not correctly handling arguments.
- #2442 Reset the "trusted" flag to false for all peers on load, before registering default hardcoded peers in the peerlist
- #26 Add additional database corruption checks in ResetCorruptDB to detect encoder ErrBufferUnderflow and ErrMaxLenExceeded

### Changed

- `type` is now a required parameter for `POST /api/v1/wallet`. `type` can be `deterministic`, `bip44` or `xpub`.
- Add `display_name`, `ticker`, `coin_hours_display_name`, `coin_hours_display_name_singular`, `coin_hours_ticker`, `explorer_url` to the `/health` endpoint response
- `cli addPrivateKey` will only work on a `collection` type wallet. Create one with `cli walletCreate -t collection`
- Don't print the wallet in the terminal after `cli encryptWallet` or `cli decryptWallet`
- Remove `WALLET_DIR` and `WALLET_NAME` envvars from the `cli` tool. Commands which need to operate on a wallet file accept the wallet file on the command line instead.
- Now the modal window for showing QR codes in the GUI allows to request specific amounts of coins, as in mobile wallets. This changes did no include the ability to read QR codes or URLs.

### Removed

- Remove `-arbitrating` option from the daemon CLI options
- Remove `-print-web-interface-address` option from the daemon CLI options
- Remove `cli walletDir` command

## [0.26.0] - 2019-05-21

### Added

- When sending coins in the UI, the user can choose to send in SKY, or the equivalent amount of SKY in USD
- Add the option for changing the language of the GUI.
- Add Spanish and Simplified Chinese language options
- Add genesis block hash in `INTR` message
- Add `bip32` package for preliminary HD wallet support
- Add CLI `checkDBDecoding` command to verify the `skyencoder`-generated binary decoders match the reflect-based decoder
- Add CLI `addresscount` command to return the count of addresses that currently have unspent outputs (coins) associated with them.
- Add `-max-inc-msg-len` and `-max-out-msg-len` options to control the size of incoming and outgoing wire messages

src/fiber/config.go
/*
Package fiber provides configuration definitions and utilities for managing fiber coins
*/
package fiber

import (
	"fmt"
	"strings"

	"github.com/spf13/viper"

	"github.com/SkycoinProject/skycoin/src/cipher/bip44"
)

// Config records fiber coin parameters
type Config struct {
	Node   NodeConfig   `mapstructure:"node"`
	Params ParamsConfig `mapstructure:"params"`
}

// NodeConfig configures the default CLI options for the skycoin node.
// These parameters are loaded via cmd/skycoin/skycoin.go into src/SkycoinProject/skycoin.go.
type NodeConfig struct {
	// Port is the default port that the wire protocol communicates over
	Port int `mapstructure:"port"`
	// WebInterfacePort is the default port that the web/gui interface serves on
	WebInterfacePort int `mapstructure:"web_interface_port"`
	// GenesisSignatureStr is a hex-encoded signature of the genesis block input
	GenesisSignatureStr string `mapstructure:"genesis_signature_str"`
	// GenesisAddressStr is the skycoin address that the genesis coins were sent to in the genesis block
	GenesisAddressStr string `mapstructure:"genesis_address_str"`
	// BlockchainPubkeyStr is a hex-encoded public key used to validate published blocks
	BlockchainPubkeyStr string `mapstructure:"blockchain_pubkey_str"`
	// BlockchainSeckey is a hex-encoded secret key required for block publishing.
	// It must correspond to BlockchainPubkeyStr
	BlockchainSeckeyStr string `mapstructure:"blockchain_seckey_str"`
	// GenesisTimestamp is the timestamp of the genesis block
	GenesisTimestamp uint64 `mapstructure:"genesis_timestamp"`
	// GenesisCoinVolume is the total number of coins in the genesis block
	GenesisCoinVolume uint64 `mapstructure:"genesis_coin_volume"`
	// DefaultConnections are the default "trusted" connections a node will try to connect to for bootstrapping
	DefaultConnections []string `mapstructure:"default_connections"`
	// PeerlistURL is a URL pointing to a newline-separated list of ip:ports that are used for bootstrapping (but they are not "trusted")
	PeerListURL string `mapstructure:"peer_list_url"`

	// UnconfirmedBurnFactor is the burn factor to apply when verifying unconfirmed transactions
	UnconfirmedBurnFactor uint32 `mapstructure:"unconfirmed_burn_factor"`
	// UnconfirmedMaxTransactionSize is the maximum size of an unconfirmed transaction
	UnconfirmedMaxTransactionSize uint32 `mapstructure:"unconfirmed_max_transaction_size"`
	// UnconfirmedMaxDropletPrecision is the maximum number of decimals allowed in an unconfirmed transaction
	UnconfirmedMaxDropletPrecision uint8 `mapstructure:"unconfirmed_max_decimals"`
	// CreateBlockBurnFactor is the burn factor to apply to transactions when publishing blocks
	CreateBlockBurnFactor uint32 `mapstructure:"create_block_burn_factor"`
	// CreateBlockMaxTransactionSize is the maximum size of an transaction when publishing blocks
	CreateBlockMaxTransactionSize uint32 `mapstructure:"create_block_max_transaction_size"`
	// CreateBlockMaxDropletPrecision is the maximum number of decimals allowed in a transaction when publishing blocks
	CreateBlockMaxDropletPrecision uint8 `mapstructure:"create_block_max_decimals"`
	// MaxBlockTransactionsSize is the maximum total size of transactions in a block when publishing a block
	MaxBlockTransactionsSize uint32 `mapstructure:"max_block_transactions_size"`

	// DisplayName is the display name of the coin in the wallet e.g. Skycoin
	DisplayName string `mapstructure:"display_name"`
	// Ticker is the coin's price ticker, e.g. SKY
	Ticker string `mapstructure:"ticker"`
	// CoinHoursName is the name of the coinhour asset type, e.g. Coin Hours
	CoinHoursName string `mapstructure:"coin_hours_display_name"`
	// CoinHoursNameSingular is the singular form of the name of the coinhour asset type, e.g. Coin Hour
	CoinHoursNameSingular string `mapstructure:"coin_hours_display_name_singular"`
	// CoinHoursTicker is the name of the coinhour asset type's price ticker, e.g. SCH (Skycoin Coin Hours)
	CoinHoursTicker string `mapstructure:"coin_hours_ticker"`
	// ExplorerURL is the URL of the public explorer
	ExplorerURL string `mapstructure:"explorer_url"`
	// VersionURL is the URL for wallet to check the latest version number
	VersionURL string `mapstructure:"version_url"`
	// Bip44Coin is the default "coin" value of the bip44 path
	Bip44Coin bip44.CoinType `mapstructure:"bip44_coin"`

	// These fields are set by cmd/newcoin and are not configured in the fiber.toml file
	CoinName      string
	DataDirectory string
}

// ParamsConfig are the parameters used to generate params/params.go.
// These parameters are exposed in an importable package `params` because they
// may need to be imported by libraries that would not know the node's configured CLI options.
type ParamsConfig struct {
	// MaxCoinSupply is the maximum supply of coins
	MaxCoinSupply uint64 `mapstructure:"max_coin_supply"`
	// InitialUnlockedCount is the initial number of unlocked addresses
	InitialUnlockedCount uint64 `mapstructure:"initial_unlocked_count"`
	// UnlockAddressRate is the number of addresses to unlock per unlock time interval
	UnlockAddressRate uint64 `mapstructure:"unlock_address_rate"`
	// UnlockTimeInterval is the distribution address unlock time interval, measured in seconds.
	// Once the InitialUnlockedCount is exhausted, UnlockAddressRate addresses will be unlocked per UnlockTimeInterval
	UnlockTimeInterval uint64 `mapstructure:"unlock_time_interval"`
	// UserMaxDropletPrecision represents the decimal precision of droplets
	UserMaxDropletPrecision uint64 `mapstructure:"user_max_decimals"`
	// UserMaxTransactionSize is max size of a user-created transaction (typically equal to the max size of a block)
	UserMaxTransactionSize int `mapstructure:"user_max_transaction_size"`
	// DistributionAddresses are addresses that received coins from the genesis address in the first block,
	// used to calculate current and max supply and do distribution timelocking
	DistributionAddresses []string `mapstructure:"distribution_addresses"`
	// UserBurnFactor inverse fraction of coinhours that must be burned, this value is used when creating transactions
	UserBurnFactor uint64 `mapstructure:"user_burn_factor"`
}

// NewConfig loads blockchain config parameters from a config file
// default file is: fiber.toml in the project root
// JSON, toml or yaml file can be used (toml preferred).
func NewConfig(configName, appDir string) (Config, error) {
	// set viper parameters
	// check that file is of supported type
	confNameSplit := strings.Split(configName, ".")
	fileType := confNameSplit[len(confNameSplit)-1]
	switch fileType {
	case "toml", "json", "yaml", "yml":
		viper.SetConfigType(confNameSplit[len(confNameSplit)-1])
	default:
		return Config{}, fmt.Errorf("invalid blockchain config file type: %s", fileType)
	}

	configName = configName[:len(configName)-(len(fileType)+1)]
	viper.SetConfigName(configName)

	viper.AddConfigPath(appDir)
	viper.AddConfigPath(".")

	// set defaults
	setDefaults()

	params := Config{}

	if err := viper.ReadInConfig(); err != nil {
		return params, err
	}

	if err := viper.Unmarshal(&params); err != nil {
		return params, err
	}

	return params, nil
}

func setDefaults() {
	// node defaults
	viper.SetDefault("node.genesis_coin_volume", 100e12)
	viper.SetDefault("node.port", 6000)
	viper.SetDefault("node.web_interface_port", 6420)
	viper.SetDefault("node.unconfirmed_burn_factor", 100)
	viper.SetDefault("node.unconfirmed_max_transaction_size", 32*1024)
	viper.SetDefault("node.unconfirmed_max_decimals", 3)
	viper.SetDefault("node.create_block_burn_factor", 100)
	viper.SetDefault("node.create_block_max_transaction_size", 32*1024)
	viper.SetDefault("node.create_block_max_decimals", 3)
	viper.SetDefault("node.max_block_transactions_size", 32*1024)
	viper.SetDefault("node.display_name", "Skycoin")
	viper.SetDefault("node.ticker", "SKY")
	viper.SetDefault("node.coin_hours_display_name", "Coin Hours")
	viper.SetDefault("node.coin_hours_display_name_singular", "Coin Hour")
	viper.SetDefault("node.coin_hours_ticker", "SCH")
	viper.SetDefault("node.explorer_url", "https://explorer.skycoin.com")
	viper.SetDefault("node.version_url", "https://version.skycoin.com/skycoin/version.txt")
	viper.SetDefault("node.bip44_coin", bip44.CoinTypeSkycoin)

	// build defaults
	viper.SetDefault("build.commit", "")
	viper.SetDefault("build.branch", "")

	// params defaults
	viper.SetDefault("params.max_coin_supply", 1e8)
	viper.SetDefault("params.initial_unlocked_count", 25)
	viper.SetDefault("params.unlock_address_rate", 5)
	viper.SetDefault("params.unlock_time_interval", 60*60*24*365)
	viper.SetDefault("params.user_max_decimals", 3)
	viper.SetDefault("params.user_burn_factor", 100)
	viper.SetDefault("params.user_max_transaction_size", 32*1024)
}
